src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,290 @@
+        this.memoryAllocator = this.configuration.MemoryAllocator;
+    }
+
+    public DecoderOptions Options { get; }
+
+    public Size Dimensions { get; }
+
+    /// <inheritdoc />
+    public Image<TPixel> Decode<TPixel>(BufferedReadStream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        // Process the header to get metadata
+        this.ProcessHeader(stream);
+
+        // Create Image object
+        ImageMetadata metadata = new()
+        {
+            DecodedImageFormat = QoiFormat.Instance,
+            HorizontalResolution = this.header.Width,
+            VerticalResolution = this.header.Height,
+            ResolutionUnits = PixelResolutionUnit.AspectRatio
+        };
+        Image<TPixel> image = new(this.configuration, (int)this.header.Width, (int)this.header.Height, metadata);
+        Buffer2D<TPixel> pixels = image.GetRootFramePixelBuffer();
+
+        this.ProcessPixels(stream, pixels);
+
+        return image;
+    }
+
+    /// <inheritdoc />
+    public ImageInfo Identify(BufferedReadStream stream, CancellationToken cancellationToken)
+    {
+        this.ProcessHeader(stream);
+        PixelTypeInfo pixelType = new(8 * (int)this.header.Channels);
+        Size size = new((int)this.header.Width, (int)this.header.Height);
+
+        ImageMetadata metadata = new();
+        QoiMetadata qoiMetadata = metadata.GetQoiMetadata();
+        qoiMetadata.Channels = this.header.Channels;
+        qoiMetadata.ColorSpace = this.header.ColorSpace;
+
+        return new ImageInfo(pixelType, size, metadata);
+    }
+
+    /// <summary>
+    /// Processes the 14-byte header to validate the image and save the metadata
+    /// in <see cref="header"/>
+    /// </summary>
+    /// <param name="stream">The stream where the bytes are being read</param>
+    /// <exception cref="InvalidImageContentException">If the stream doesn't store a qoi image</exception>
+    private void ProcessHeader(BufferedReadStream stream)
+    {
+        Span<byte> magicBytes = stackalloc byte[4];
+        Span<byte> widthBytes = stackalloc byte[4];
+        Span<byte> heightBytes = stackalloc byte[4];
+
+        // Read magic bytes
+        int read = stream.Read(magicBytes);
+        if (read != 4 || !magicBytes.SequenceEqual(QoiConstants.Magic.ToArray()))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // If it's a qoi image, read the rest of properties
+        read = stream.Read(widthBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        read = stream.Read(heightBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // These numbers are in Big Endian so we have to reverse them to get the real number
+        uint width = BinaryPrimitives.ReadUInt32BigEndian(widthBytes),
+            height = BinaryPrimitives.ReadUInt32BigEndian(heightBytes);
```

Comment:
I guess there is no rule for this in StyleCop , but we don't use the multi-assignment in declarations.

```suggestion
        uint width = BinaryPrimitives.ReadUInt32BigEndian(widthBytes);
        uint height = BinaryPrimitives.ReadUInt32BigEndian(heightBytes);
```